src/hotspot/share/code/aotCodeCache.cpp

```diff
@@ -0,0 +1,1283 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "cds/aotCacheAccess.hpp"
+#include "cds/cds_globals.hpp"
+#include "cds/cdsConfig.hpp"
+#include "cds/heapShared.hpp"
+#include "cds/metaspaceShared.hpp"
+#include "classfile/javaAssertions.hpp"
+#include "code/aotCodeCache.hpp"
+#include "code/codeCache.hpp"
+#include "gc/shared/gcConfig.hpp"
+#include "logging/logStream.hpp"
+#include "memory/memoryReserver.hpp"
+#include "runtime/flags/flagSetting.hpp"
+#include "runtime/globals_extension.hpp"
+#include "runtime/java.hpp"
+#include "runtime/os.inline.hpp"
+#include "runtime/sharedRuntime.hpp"
+#include "runtime/stubRoutines.hpp"
+#ifdef COMPILER2
+#include "opto/runtime.hpp"
+#endif
+#if INCLUDE_G1GC
+#include "gc/g1/g1BarrierSetRuntime.hpp"
+#endif
+#if INCLUDE_ZGC
+#include "gc/z/zBarrierSetRuntime.hpp"
+#endif
+
+#include <sys/stat.h>
+#include <errno.h>
+
+static void exit_vm_on_load_failure() {
+  if (AbortVMOnAOTCodeFailure) {
+    vm_exit_during_initialization("Unable to use AOT Code Cache.", nullptr);
+  }
+  AOTAdapterCaching = false;
+}
+
+static void exit_vm_on_store_failure() {
```

Comment:
It's a bit confusing to see `exit_vm_on_load_failure()` and `exit_vm_on_store_failure()` to silently proceed unless a flag is explicitly specified.

Moreover, how reliable `AOTAdapterCaching = false` to fail-fast and avoid repreated load/store attempts? At least, I see that `AOTCodeCache` ctor cache `AOTAdapterCaching`, so it won't see the update.

How does it affect adapter code generation during assembly phase?

Reply:
We check failure state of AOT code cache when query about using adapters caching:
```
  bool for_use()  const { return _for_use  && !_failed; }
  bool for_dump() const { return _for_dump && !_failed; }
  static bool is_on() CDS_ONLY({ return _cache != nullptr && !_cache->closing(); }) NOT_CDS_RETURN_(false);
  static bool is_on_for_use()  { return is_on() && _cache->for_use(); }
  static bool is_on_for_dump() { return is_on() && _cache->for_dump(); }
```

```
  static bool is_dumping_adapters() { return is_on_for_dump() && _cache->adapter_caching(); }
  static bool is_using_adapters()   { return is_on_for_use() && _cache->adapter_caching(); }
```

Reply:
AOT adapters code caching is loading is guarded by the methods not by flag.

Setting AOTAdapterCaching to false on failure is simple indication that adapter caching is switched of for someone who will look on final state of flag.

Reply:
I added `log_info()` to `exit_vm_on_*_failure()` methods to produce notification when AbortVMOnAOTCodeFailure flag is off (default value).

Reply:
> Setting AOTAdapterCaching to false on failure is simple indication that adapter caching is switched off for someone who will look on final state of flag.

But how does it affect execution? What's the intended behavior when a failure happens during store attempt? What are the consistency guarantees for AOT code cache during dumping in presense of store failures? What I see right now is that errors reported by `AOTCodeCache::store_code_blob()` are silently ignored. How does `_cache` notice a store failure during dumping phase?